### ISSUE TITLE
Extra placeholders for missing modded content

### ISSUE THREAD

**Issue by Mach12** [tModLoader/tModLoader#37]
I know that very, VERY important parts of modded content are already placeholder'd when missing, and restored when the mod comes back. I just thought I'd throw out there a note of things one still loses upon disabling and reenabling a mod:
- Prefixes on items, except (for some reason) accessories in 1-6th slots. Accessories in modded slots still lose their prefixes
- Walls
- Modded armor on mannequins
- Townfolk (yes they come back slowly, but when you test things on vanilla and come back to a whole modpack, it can take several in-game days to get them all back)

So... yeah. Here's a list for reference. I guess it's not major, but the more we keep content when temporarily removing a mod, the better amirite?

**Comment by Chicken-Bones** [comment]
I agree with the first three, though I haven't looked into how difficult they will be. 
As for Townfolk, it may be difficult to manage things such as their rooms getting taken or destroyed. Perhaps it's better to add a console command to fill all available houses with valid town NPCs

**Comment by Mach12** [comment]
How about making them swarm back in when resuming the game? If it's hard to make them seem to always have been there, maybe we could dial such command or similar when joining a world that has lost town NPCs...

**Comment by bluemagic123** [comment]
Hm, I could think of a way to make prefixes, modded armor on mannequins, and townfolk stay. However, walls would be pretty difficult to preserve. I actually wanted to make that happen in the most recent update, but because literally the only thing saved about walls is their type, there isn't any way to preserve them without storing extra per-tile data externally.

**Comment by Chicken-Bones** [comment]
On second thoughts I'm not so sure about retaining prefixes and walls, due to save bloat. Perhaps there should instead be a confirmation message when loading a player/world when such things would be lost

**Comment by Mach12** [comment]
I'm not sure about save bloating for prefixes, but maybe the wall data would be compressible?
I must admit that I'm in half-unknown territory, but travelling around in a laggy server showed me that the map is scattered into chunks.
If it's also the case in the save, maybe we could just not save chunks that do not contain modded walls?
Supposing that no mod replaces every wall in the world, this could spare us most/all of the sky, same for hell, probably also the jungle and dungeon... actually, most chunks would probably be just builded areas and randomly generated underground structures...

**Comment by bluemagic123** [comment]
The map isn't scattered into chunks; it's actually loaded all at once, and it's saved using for loops to iterate over the entire thing. I'm not really worried about save bloating, but the main problem with unloaded walls is that either an entirely new array representing the map will need to be allocated just to store unloaded wall information, or the runtime memory usage of Tile will need to be increased just for unloaded wall information.

I don't think save bloating with prefixes would be too much of a problem, considering items don't take up too much relative memory and it would only be stored for unloaded items (which wouldn't happen too much anyways).

**Comment by Chicken-Bones** [comment]
Ok, I retract my previous statement, placeholder walls/prefixes seem like a good idea if we can figure out how to implement them

**Comment by JavidPack** [comment]
Unloaded Chests are lost forever still, as are Walls, we still have things to fix here.

**Comment by Solxanich** [comment]
I'm gonna tackle this. 
Some additional considerations after working on it a bit:
1. Preserving non-solid vs. solid blocks so as to minimize impacts on liquid preservation and player movement?
2. Chests should or shouldn't preserve Locked Status and related Locking Item?
3. Do we also want to implement the code for making it possible to placeholder modded Liquids (since implementing placeholder walls is going to be very similar)?

**Comment by direwolf420** [comment]
Just got pointed towards your comment here. I was starting on making an "UnloadedWall" implementation three days ago and got it into a functional state today: [image 1](https://cdn.discordapp.com/attachments/695301209170509885/793100479965954088/Screenshot_6.png), [image 2](https://cdn.discordapp.com/attachments/695301209170509885/793100481797816350/Screenshot_7.png). In the process I also documented the now 5 year old code handling unloaded tiles, so these would be helpful later.

There are ways to clean up the current system which would make implementing your points easier, which I will get to later after I clean up my own code and do a PR (either inside the PR or a separate one). Stay tuned.

**Comment by Solxanich** [comment]
RIP. Unfortunate, as I had also only started two days ago. Although, maybe luckily, I wasn't working on UnloadedWall, as I had focused on UnloadedChest first. Depending on what you've done on that front, I may be able to merge my work in after.

I've implemented the following in regards to UnloadedChests:
1. Documenting sections of the 5 year old code - likely this will be redundant
2. Implemented NonSolid Unloaded Block - fully implemented and finished testing
3. 90% implemented UnloadedChest, exempting locked related data, probably would finish in next 5hrs.

If there is no overlap on unloadedChests, I'll try and get a branch up later today, if not, let me know to stop @direwolf420 .

**Comment by direwolf420** [comment]
See the last three commits here for a very rudimentary, non-refactored/cleaned up draft for UnloadedWall. https://github.com/direwolf420/tModLoader/commits/1.4_modwall_saving_sample_code
@Solxanich see if my documentation lines up with yours. I know its not complete but I just commented stuff that helped me understand, and I didn't comment things in Read code if they were commented in Write

**Comment by Solxanich** [comment]
Thanks direwolf420.
I've merged in your changes to my own branch and resolved all conflicts; both sets of changes run together. I still have work to do on unloadedChests, as there is a few issues with my current implementation.
[https://github.com/Solxanich/tModLoader/tree/1.4_UnloadedChest](https://github.com/Solxanich/tModLoader/tree/1.4_UnloadedChest)
Documentation mostly lines up, with us adding info to different areas, the merged branch combines all added comments accordingly. 

Both branches have the same memory issue of adding more storage; I agree there are better ways to do this, but also agree its a future discussion compared to patching for now.

**Comment by Chicken-Bones** [comment]
With unloaded content, the primary goal is for people to realise the forgot a mod, and reenable it, without data loss. The secondary goal is to allow them to remove that content destructively. Improving playability with unloaded content is not a necessity.

1. Solid vs non solid blocks sound like a good idea
2. Chest locking is not required, I'm happy for items in chests to he inaccessible when the chest is unloaded.
3. Unloaded liquid should wait for ModLiquid implementation
4. I haven't looked at the code, but there's a good chance unloaded wall data can be compressed significantly, I'd suggest a chunk system, combined with run length encoding.

**Comment by direwolf420** [comment]
@Solxanich I checked your branch and I can see it turning into a mess already, theres things that would help organizing this for the current stuff in tml + what me and you have. We could use your branch as a "playground" to test the new features and how they interact together (there's gonna be many world-breaking changes I bet).

Just spitballing here so don't take this as a very thought out proposal.

Splitting up this:
```cs
string modName = infoTag.GetString("mod");
string name = infoTag.GetString("name");
bool IsChest = infoTag.GetBool("IsChest");
bool IsSolid = infoTag.GetBool("IsSolid");
short ChestIndx = infoTag.GetShort("ChestIndx");
bool frameImportant = infoTag.ContainsKey("frameX");
```

Done through OOPing the current `UnloadedTileInfo` into a base object that just holds "mod" and "name" values, and then inheriting ones for tiles: +isSolid, +frameX/Y. Walls would not have any additional info. Chests would use one that adds more chest related info. This would clarify the code by not overloading a single class and its constructor, and narrow down its specific purpose. Plus, this general base object could be used for many other things ("mod"+"name" is seen in Item IO too).

Ultimately we'd track these objects separately (TileInfo, WallInfo, ChestInfo, LiquidInfo). Each one has an associated unloaded and pending unloaded type.

Another thing that would make saving a tad bit more efficient is creating a hierarchy (in TagCompound form atleast):
Currently:
```
{
  { "ExampleMod" : "ExampleTile" }, 
  { "ExampleMod" : "ExampleWorkbench" }, 
  { "ExampleMod" : "ExampleBed" }, 
  { "WhateverMod" : "Whatever" }
}
```

Possible option:
```
{
  { "ExampleMod" : ["ExampleTile", "ExampleWorkbench", "ExampleBed"] }
  { "WhateverMod" : ["Whatever"] }
}
```

Lastly, one thing during development is problematic and that is modifying the data in the Write/Read (for example you adding a second `flags` variable, increasing the length of the data array) without making worlds necessarily backwards compatible somewhere down the line (I think the goal is to make 1.3 worlds loadable in 1.4). Sadly the current `flags` variable is already full (uses all 8 bits), so I'm not sure how to proceed further without adding a new variable.

Also yes @Chicken-Bones, nothing has been done to compress any data as of now, and I didn't really expect you to point it out at this stage of development.

**Comment by Solxanich** [comment]
@direwolf420 
Agreed on the OOP refactor - the branch is really messy as is, with multiple different save/load styles and extraneous information abound. Liquids would only worsen the problem as is. 
I'll take a look at refactoring that.
I can see the point on hierarchy - if we already know the mod is unloaded etc. There is carry-down information. I'll take a look after refactoring.

Some further thoughts on what I'd like to also change:
1 ) Swap from using frameX,Y data to index tiles, and instead use coordsX,Y as the style. This would then create support for walls, tiles, liquids, and chests properly under the same save style. 
Current: `int FrameID = frameY * (short.MaxValue) + frameX` would be swapped for ` uint32 posID = Y * ( Main.maxTilesX) + X`. This would be universal index to all, and eliminate loading errors I'm seeing on chests as chest frames change.
2 ) Move the previously written SaveChests/LoadChests in worldIO to be taken care of within the refactored implementation
3 ) Try to rework the flags so that it works with older worlds and would support 16 bits over 8, as liquids and chests pushes it to 10+ bits . Probably safest would be to` if(oldVersion) then flags2 = 0 else read flags2`, but I'd need to look into it.

**Comment by Solxanich** [comment]
@direwolf420 
The following branch contains a partially-refactored (I haven't changed tile frame based loading or touched walls) version with a working basic implementation of placeholder Chests and placeholder Non-solid tiles.
Thoughts on if its cleaner to the level we're looking for? 
I have temporarily removed the extra flags, for now, as that isn't necessary for basic implementation, but likely will still come into play in the future with liquids and unmanaged chest data.

**Comment by direwolf420** [comment]
yep, yep, definitely goes into the right direction. There can be many improvements made still since this is obviously more of an experiment (clean up constructors, add deconstructor, make more methods virtual instead of new). As for chests, I can't say much. As for non-solid tiles, those could get OOPed aswell (their only difference is two lines of code in SetStaticDefaults, and in MouseOver).

Great work so far, I would suggest that you write up possible improvements to the current implementation here after this "experiment" is concluded, to make it easier on everyone later to understand/review it, aswell as make it more open to contributions down the line. 

As an aside, I would like it if this problem would be tackled first by identifying improvements for the **current implementation**, then fixing them, and then start one-by-one adding unloaded things (non-solid tiles, walls, then chests). I would worry about compressing data later when the implementation is set in stone.

I'm personally not fused to my current progress, happy to redo my branch after the improvements are made.

**Comment by Solxanich** [comment]
@direwolf420 - I've mostly finished refactoring the branch, and everything runs well. Walls, tiles, chests all follow the same implementation, and have been refactored down to eliminate some redundant code (example: the majority of the pending implementation). I believe it should

If you're content with the state of walls as updated, I'll look to create a PR with our changes and get the basic implementation to review.

Future improvements for future separate PRs:
1) Expand TileIO's ReadModTile and WriteModTile to encompass 16 bits of flags, so that liquid amount and 'chest style'/'lockedchest was unlocked' can be preserved. Must also address backwards compatibility of this enhancement within same PR.
2) Further re-factor boiler plate code in UnloadedTilesWorld, in particular in the load function.
3) Add modded liquids, modded wires, etc. as requested based on modded walls implementation

Concerns with reworked branch:
The unloaded rework will not be backwards compatible with worlds containing placeholder objects.

Sample Images
Before:
![tML_placeholder_before](https://user-images.githubusercontent.com/59670736/103448523-ec953800-4c57-11eb-8b1b-883750a49a02.png)

During:
![tML_placeholder_during](https://user-images.githubusercontent.com/59670736/103448525-f5860980-4c57-11eb-8ebb-0187dd5b93e9.png)

After:
![tML_placeholder_after](https://user-images.githubusercontent.com/59670736/103448529-03d42580-4c58-11eb-88c1-2d9670b770a8.png)

**Comment by direwolf420** [comment]
I'm worried about the `tileInfoMap` you created, before it would be stored on the tiles frameX/Y to save space, now it saves separately, creating alot of data. The implementation and purpose of `workingType` is a bit iffy aswell imo.

As I said in my previous post, I would like this whole process to be step-by-step, instead of all at once like your experimental PR, as I got kinda lost in it (even though I followed all the discussions).

Though I got important insights from your code, which is an important step towards the final implementation. Someone more knowledgeable needs to figure out a way to make worlds backwards compatible when this all is done. (Implementation and expansion of `UnloadedPosIndexing` is the crucial part here I think, which should be expanded to store any type of data efficiently on a tile, usable by other mods)

(Unrelated but undo your `Terraria.csproj.patch` change, and simply add the textures into the Default folder both in `src` aswell as `patches`, then "diff tmodloader" in the setup tool, I had a similar issue with them not being recognized by source control)

**Comment by bvcbvc** [comment]
wall could be black and pink maybe
